scripts/patch_websockets.py

```diff
@@ -0,0 +1,91 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Patches uWebsockets Socket.h to fix a crash sending large
+payloads with ssl encoding.
+"""
+from absl import app
+from absl import flags
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_string("file", None, "Sockets.h file to patch.")
+
+# The line of Socket.h that we intend to start overwriting, and the 
+# number of lines to overwite.
+REPLACE_BLOCK_START_LINE = 308
+REPLACE_BLOCK_OVERWRITE_LENGTH = 17
+
+# The code which fixes the crash which will replace the existing
+# error handling implementation with one that retries the writes
+# when the SSL encoding buffer is full.
+REPLACE_BLOCK_CODE_LINES = [
+  "              bool continue_loop = true;\n",
+  "              do {\n",
+  "                sent = SSL_write(ssl, message->data, message->length);\n",
+  "                static int bytes_written = 0;\n",
+  "                bytes_written += sent;\n",
+  "                if (sent == (ssize_t) message->length) {\n",
+  "                  wasTransferred = false;\n",
+  "                  return true;\n",
+  "                }\n",
+  "                else if (sent < 0) {\n",
+  "                  switch (SSL_get_error(ssl, sent)) {\n",
+  "                  case SSL_ERROR_WANT_READ:\n",
+  "                    continue_loop = false;\n",
+  "                    break;\n",
+  "                  case SSL_ERROR_WANT_WRITE:\n",
+  "                    break;\n",
+  "                  default:\n",
+  "                    return false;\n",
+  "                }\n",
+  "              }\n",
+  "            }  while (continue_loop);\n"
```

Comment:
Is it possible that SSL_ERROR_WANT_WRITE occurs forever and this loop never ends?
Perhaps we can set a retry limit like 100 times?

Reply:
I tried doing just that. First off, I found it fuzzy to estimate a cap on the number of iterations of the loop, since the number of iterations will depend on indeterminate factors like encoding time, network usage, etc.  The loop count varied wildly. Addtionally returning false from this function to signify an error causes the uWebsockets implementation to hang, so even if we decided to opt out of the write then their library doesn't handle it well.

At any rate, an idefinite loop  shouldn't occur.  This is a buffer consumption issue where the underlying SSL process funnels the data out of the full SSL buffer and into the socket.  Eventually. it will empty unless there's an issue with the socket itself (timeout, broken pipe, etc) and  and the SSL_write operation will return an error other than WANT_WRITE.  Of course, the uWebsockets implementation then hangs indefinitely again, which is the same thing as an infinite loop anyway.

Generally speaking this version of uWebSockets doesn't seem very robust. It obviously. doesn't handle errors well. I think of this PR as a simpy a temproary patch of buggy code to unblock our users.  Really we should replace the entire implementation, but that's a signifcant amount of work that. should be scheduled.